types/src/v0/impls/stake_table.rs

```diff
@@ -26,84 +37,217 @@ use hotshot_types::{
     }
 
-    /// Create the consensus and DA stake tables from L1 events
-    ///
-    /// This is a pure function, to make it easily testable.
-    ///
-    /// We expect have at most a few hundred EVM events in the
-    /// PermissionedStakeTable contract over the liftetime of the contract so it
-    /// should not significantly affect performance to fetch all events and
-    /// perform the computation in this functions once per epoch.
-    pub fn from_l1_events(updates: Vec<StakersUpdated>) -> Self {
-        let changes_per_node = updates
-            .into_iter()
-            .flat_map(|event| {
-                event
-                    .removed
-                    .into_iter()
-                    .map(|key| StakeTableChange::Remove(bls_alloy_to_jf(key)))
-                    .chain(
-                        event
-                            .added
-                            .into_iter()
-                            .map(|node_info| StakeTableChange::Add(node_info.into())),
-                    )
-            })
-            .group_by(|change| change.key());
+    for (dereg, log) in deregistered {
+        st_map.insert(
+            (
+                log.block_number.unwrap(),
+                log.transaction_index.unwrap(),
+                log.log_index.unwrap(),
+            ),
+            StakeTableChange::Remove(dereg),
+        );
+    }
 
-        // If the last event for a stakers is `Added` the staker is currently
-        // staking, if the last event is removed or (or the staker is not present)
-        // they are not staking.
-        let currently_staking = changes_per_node
-            .into_iter()
-            .map(|(_pub_key, deltas)| deltas.last().expect("deltas non-empty").clone())
-            .filter_map(|change| match change {
-                StakeTableChange::Add(node_info) => Some(node_info),
-                StakeTableChange::Remove(_) => None,
-            });
-
-        let mut consensus_stake_table: Vec<PeerConfig<PubKey>> = vec![];
-        let mut da_members: Vec<PeerConfig<PubKey>> = vec![];
-        for node in currently_staking {
-            consensus_stake_table.push(node.clone().into());
-            if node.da {
-                da_members.push(node.into());
-            }
+    let mut delegator_map = BTreeMap::new();
+    for (del, log) in delegated {
+        delegator_map.insert(
+            (
+                log.block_number.unwrap(),
+                log.transaction_index.unwrap(),
+                log.log_index.unwrap(),
+            ),
+            DelegationChange::Add(del),
+        );
+    }
+
+    for (undelg, log) in undelegated {
+        delegator_map.insert(
+            (
+                log.block_number.unwrap(),
+                log.transaction_index.unwrap(),
+                log.log_index.unwrap(),
+            ),
+            DelegationChange::Remove(undelg),
+        );
+    }
+
+    let mut validators = IndexMap::new();
+    for staker in st_map.values() {
```

Comment:
@imabdulbasit do I understand correctly that we first process all the validator registrations an de-registrations and then the delegations? That seems incorrect to me. I think it's easier to reason about correctness if we process all the events in order, for example by making a an enum that can hold all the events, then sorting that by (block, log_index) and iterating over all the events and process them.

Reply:
oh that there could be multiple events for (block, log_index)?

Reply:
So, my thinking was we only need to process delegation events for validators that exist?

Reply:
So, the (block, log_index) would make each event unique, right? I can make the enum change

Reply:
https://github.com/EspressoSystems/espresso-network/pull/2772/commits/83f18136918149a6f57275631f212ef3b93e06c1